VersionControlUI/src/commands/CDiff.cpp

```diff
@@ -45,50 +40,152 @@ using namespace FilePersistence;
+		if (!name().startsWith(commandTokensCopy.takeFirst()))
+			return false;
+
+		// if there is an additional version specified, check that it is a valid commit
+		if (!commandTokensCopy.isEmpty() && !repository.isValidRevisionString(commandTokensCopy.takeFirst()))
+			return false;
+
+		// check the same if there is a second version specified
+		if (!commandTokensCopy.isEmpty() && !repository.isValidRevisionString(commandTokensCopy.takeFirst()))
+			return false;
+
+		return true;
+	}
+	else return false;
+}
+
+Interaction::CommandResult* CDiff::execute(Visualization::Item*, Visualization::Item* target,
+				const QStringList& commandTokens, const std::unique_ptr<Visualization::Cursor>& )
 {
 	auto scene = target->scene();
 	scene->clearFocus();
 	scene->clearSelection();
 	scene->setMainCursor(nullptr);
 
-	Model::TreeManager* headManager = target->node()->manager();
-	QString managerName = headManager->name();
+	QString managerName = target->node()->manager()->name();
+
 
-	// TODO add functionality to specify two versions
-	VersionControlUI::DiffManager diffManager{name, "HEAD", managerName, Model::SymbolMatcher{"Class"}};
+	// TODO restrict versions to versionA always be older?
+	// TODO diff with no versions should compare HEAD to workingDir if possible
+	// try to get the versions
+	QString versionA = commandTokens.value(1, "HEAD");
+	QString versionB = commandTokens.value(2, "HEAD");
+
+	VersionControlUI::DiffManager diffManager{versionA, versionB, managerName, Model::SymbolMatcher{"Class"}};
 	diffManager.visualize();
 
 	return new Interaction::CommandResult{};
 }
 
-QStringList CDiff::possibleNames(Visualization::Item*, Visualization::Item* target,
-											const std::unique_ptr<Visualization::Cursor>&)
+QList<Interaction::CommandSuggestion*> CDiff::suggest(Visualization::Item*, Visualization::Item* target,
+		const QString& textSoFar, const std::unique_ptr<Visualization::Cursor>&)
 {
-	Model::TreeManager* headManager = target->node()->manager();
-	QString managerName = headManager->name();
+	QStringList tokensSoFar = textSoFar.split(" ");
+
+	QList<Interaction::CommandSuggestion*> suggestions;
+	QString commandName = name();
+
+	// no suggestions for that many tokens
+	if (tokensSoFar.size() > 3)
+		return {};
+
+	// check that the command name starts with the characters of the first token
+	if (name().startsWith(tokensSoFar.takeFirst()))
+	{
+		// no additional versions specified
+		if (tokensSoFar.isEmpty())
+		{
+			// TODO implement this behavior
+			suggestions.append(new Interaction::CommandSuggestion{name(), "diff working directory against head"});
+			return suggestions;
+		}
+
+		QString firstVersionToken = tokensSoFar.takeFirst();
+		QString secondVersionToken = "";
+		bool secondVersionAvailable = !tokensSoFar.isEmpty();
+		if (secondVersionAvailable) secondVersionToken = tokensSoFar.takeFirst();
+
+
+		for (auto firstTokenEntry : commitsWithDescriptionsStartingWith(firstVersionToken, target))
```

Comment:
I don't think we should have nested loops here. You should always just looup the last available token, either the first one or the second one, and offer autocompletions with the initial text + the completion of the last token.

Btw, if I understand git correctly, you can specify a partial commit id, as long as it is unambiguous. Thus it would be good if your command work with partial ids as well (I guess you just need to forward those). So for example, it would be nice if I could type `diff abcd34 3323a` and hit enter, without having to complete to the full-blown names.

A step further would be, for the auto-completion itself to only complete an id as much as needed to make it unambiguous, but not more

so given the ids `abc34567` and `abc99999`

You could offer completions that look like this: `diff abc3` and `diff abc9`